KIPs/kip-37.md

```diff
@@ -0,0 +1,908 @@
+  8. If the above call to `supportsInterface(0x7cc2d017)` on `_to` reverts or returns a value other than the constant value `true` the `myTransferFrom` function MAY consider this transfer successful.
+     - **NOTE**: this MAY result in unrecoverable tokens if sent to an address that does not expect to receive KIP-37 tokens.
+- The above example is not exhaustive but illustrates the major points (and shows that most are shared with `safeTransferFrom` and `safeBatchTransferFrom`):
+  - Balances that are updated MUST have equivalent transfer events emitted.
+  - A receiver address has to be checked if it is a contract and if so relevant `KIP37TokenReceiver` hook function(s) have to be called on it.
+  - Balances (and events associated) that are referenced in a call to an `KIP37TokenReceiver` hook MUST be updated (and emitted) before the `KIP37TokenReceiver` hook is called.
+  - The return values of the `KIP37TokenReceiver` hook functions that are called MUST be respected if they are implemented.
+  - Only non-standard transfer functions MAY allow tokens to be sent to a recipient contract that does NOT implement the necessary `KIP37TokenReceiver` hook functions. `safeTransferFrom` and `safeBatchTransferFrom` MUST revert in that case (unless it is a hybrid standards implementation see "Backwards Compatibility").
+
+##### **_Minting-creating and burning-destroying rules:_**
+
+- A mint/create operation is essentially a specialized transfer and MUST follow these rules:
+  - To broadcast the existence of a token ID with no initial balance, the contract SHOULD emit the `TransferSingle` event from `0x0` to `0x0`, with the token creator as `_operator`, and a `_value` of 0.
+  - The [TransferSingle and TransferBatch event rules](#transfersingle-and-transferbatch-event-rules) MUST be followed as appropriate for the mint(s) (i.e., singles or batches) however the `_from` argument MUST be set to `0x0` (i.e., zero address) to flag the transfer as a mint to contract observers.
+    - **NOTE:** This includes tokens that are given an initial balance in the contract. The balance of the contract MUST also be able to be determined by events alone meaning initial contract balances (for e.g., in construction) MUST emit events to reflect those balances too.
+- A burn/destroy operation is essentially a specialized transfer and MUST follow these rules:
+  - The [TransferSingle and TransferBatch event rules](#transfersingle-and-transferbatch-event-rules) MUST be followed as appropriate for the burn(s) (i.e. singles or batches) however the `_to` argument MUST be set to `0x0` (i.e. zero address) to flag the transfer as a burn to contract observers.
+  - When burning/destroying you do not have to actually transfer to `0x0` (that is impl specific), only the `_to` argument in the event MUST be set to `0x0` as above.
```

Comment:
```suggestion
  - When burning/destroying, you do not have to actually transfer to `0x0` (that is implementation specific), only the `_to` argument in the event MUST be set to `0x0` as above.
```

Reply:
Is there any reason that it is transferred to non-zero address? If the token is sent to `0x0`, there should be some possibility that contract storage can be reduced.

Reply:
'When burning/destroying you do not have to actually transfer to `0x0`' does not mean sending to non-zero address. It means deleting from contract storage without 'transfer' and only generating 'Transfer' event.